internal/rpc/rpc.go

```diff
@@ -0,0 +1,170 @@
+		return nil, fmt.Errorf("RPC_URL environment variable is not set")
+	}
+	log.Debug().Msg("Initializing RPC")
+	rpcClient, dialErr := gethRpc.Dial(rpcUrl)
+	if dialErr != nil {
+		return nil, dialErr
+	}
+
+	ethClient := ethclient.NewClient(rpcClient)
+
+	rpc := &Client{
+		RPCClient:        rpcClient,
+		EthClient:        ethClient,
+		URL:              rpcUrl,
+		IsWebsocket:      strings.HasPrefix(rpcUrl, "ws://") || strings.HasPrefix(rpcUrl, "wss://"),
+		BlocksPerRequest: GetBlockPerRequestConfig(),
+	}
+	checkErr := rpc.checkSupportedMethods()
+	if checkErr != nil {
+		return nil, checkErr
+	}
+
+	chainIdErr := rpc.setChainID()
+	if chainIdErr != nil {
+		return nil, chainIdErr
+	}
+	return rpc, nil
+}
+
+func (rpc *Client) Close() {
+	rpc.RPCClient.Close()
+	rpc.EthClient.Close()
+}
+
+func (rpc *Client) checkSupportedMethods() error {
+	var blockByNumberResult interface{}
+	err := rpc.RPCClient.Call(&blockByNumberResult, "eth_getBlockByNumber", "latest", true)
+	if err != nil {
+		return fmt.Errorf("eth_getBlockByNumber method not supported: %v", err)
+	}
+	log.Debug().Msg("eth_getBlockByNumber method supported")
+
+	var getLogsResult interface{}
+	logsErr := rpc.RPCClient.Call(&getLogsResult, "eth_getLogs", map[string]string{"fromBlock": "0x0", "toBlock": "0x0"})
+	if logsErr != nil {
+		return fmt.Errorf("eth_getLogs method not supported: %v", logsErr)
+	}
+	log.Debug().Msg("eth_getLogs method supported")
+
+	var traceBlockResult interface{}
+	if config.Cfg.RPC.Traces.Enabled {
+		if traceBlockErr := rpc.RPCClient.Call(&traceBlockResult, "trace_block", "latest"); traceBlockErr != nil {
+			log.Warn().Err(traceBlockErr).Msg("Optional method trace_block not supported")
+		}
+	}
+	rpc.SupportsTraceBlock = traceBlockResult != nil
+	log.Debug().Msgf("trace_block method supported: %v", rpc.SupportsTraceBlock)
+	return nil
+}
+
+func (rpc *Client) setChainID() error {
+	chainID, err := rpc.EthClient.ChainID(context.Background())
+	if err != nil {
+		return fmt.Errorf("failed to get chain ID: %v", err)
+	}
+	rpc.ChainID = chainID
+	return nil
+}
+
+func (rpc *Client) GetFullBlocks(blockNumbers []*big.Int) []GetFullBlockResult {
+	var wg sync.WaitGroup
+	var blocks []RPCFetchBatchResult[common.RawBlock]
+	var logs []RPCFetchBatchResult[common.RawLogs]
+	var traces []RPCFetchBatchResult[common.RawTraces]
+
+	wg.Add(2)
+
+	go func() {
+		defer wg.Done()
+		blocks = RPCFetchBatch[common.RawBlock](rpc, blockNumbers, "eth_getBlockByNumber", func(blockNum *big.Int) []interface{} {
+			return []interface{}{hexutil.EncodeBig(blockNum), true}
+		})
+	}()
+
+	go func() {
+		defer wg.Done()
+		logs = RPCFetchInBatches[common.RawLogs](rpc, blockNumbers, rpc.BlocksPerRequest.Logs, config.Cfg.RPC.Logs.BatchDelay, "eth_getLogs", func(blockNum *big.Int) []interface{} {
+			return []interface{}{map[string]string{"fromBlock": hexutil.EncodeBig(blockNum), "toBlock": hexutil.EncodeBig(blockNum)}}
+		})
+	}()
+
+	if rpc.SupportsTraceBlock {
+		wg.Add(1)
```

Comment:
why trace fetching is isolated into it's own waiting group?

Reply:
because it's optional

Reply:
can we make the length of the waiting group dependent on the traces flag and have trace fetching belong to the same waiting group when it's active

Reply:
Why exactly?
What's the benefit of doing

```
numOfWgs := 2
if (traces) {
  numOfWgs = 3
}
wg.add(numOfWgs)
if (traces) {
  // trace call
}
```

instead of 

```
wg.add(2)
...
if (traces) {
  wg.add(1)
  // trace call
}
```

Reply:
it's less synchronous. but Im not sure how difference it makes, so resolving this

Reply:
it's not less synchronous, the end result is the same since `wg.Wait()` is called afterwards anyways